File: Kotlin/search/StairCaseSearch.kt
// Kotlin program to implement Staircase Search using Iterative Approach

import java.util.Scanner

fun main(args: Array <String>) {

    val input = Scanner(System.`in`)

    val noOfRows = input.nextInt()
    val noOfColumns = input.nextInt()

    val matrix = ArrayList <ArrayList <Int>> ()
    for(row in 0 until noOfRows) {
        var rows = ArrayList <Int> ()
        for(column in 0 until noOfColumns) {
            var element = input.nextInt()
            rows.add(element)
        }
        matrix.add(rows)
    }

    val key = input.nextInt()
    stairCaseSearch(matrix , key)
}

fun stairCaseSearch(matrix: ArrayList<ArrayList<Int>>, key: Int) {

    var noOfRows = matrix.size
    var noOfColumns = matrix[0].size
    var row = 0
    var column = noOfColumns - 1
    var isKeyPresent = false

    while(row <= noOfRows - 1 && column >= 0) {

        // If key is equal to the current element, print the indices values and break
        if(matrix[row][column] == key) {
            isKeyPresent = true
            println("Key: $key is present at Row: ${row + 1} and Column: ${column + 1}")
            break
        }

        // If current element is greater than key, then since it is column sorted matrix, the
        // previous column would have smaller element than the current column
        else if(matrix[row][column] > key) {
            column--
        }

        // If the current element is smaller than key, then since it is row sorted matrix, the
        // next row would have greater element than the current row
        else {
            row++
        }
    }

    // If the element is not present in the matrix
    if(!isKeyPresent)
        println("Key: $key is not present in the matrix")
}

// Time Complexity: O(n + m)
// Auxiliary Space Complexity: O(1)

/* Sample Test Cases:
Test - 1:
    Input
        3 4
        1 2  3  4
        5 6  7  8
        9 10 11 12
        11

    Output
        Key: 11 is present at Row: 3 and Column: 3

Test - 2:
    Input
        2 1
        10
        20
        -10
    Output
        Key: -10 is not present in the matrix
 */

File: C/search/Staircase_Search.c
/*
Staircase Search is used to search for a key in a 2D Array whose row and column elements are sorted.
	* The idea is to remove a row or a column in each comparison until an element is found. 
	* Searching is started from the top-right corner of the matrix and has three possible cases: 
		1. The key is greater than the current element, this means, that all the elements in the current row are smaller than the key.
       So, the row can be skipped and move to next row.
		2. The key is smaller than the current number, this means, that all the elements in the current column are greater than the key.
       So, the column can be skipped and move to previous column.
		3. The key is equal to the current number, this means, the key is found.
	* If the bounds of the 2D Array are reached, this means, the key is not found.
*/

#include<stdio.h>     
#include<stdlib.h>    

void staircase_search(int** matrix, int n, int m, int key) {

	int is_found=0;
	// Start from top right corner
	int i = 0, j = m - 1;

	// To search for a element until the indices satisfy the matrix bounds
	while (i <= n - 1 && j >= 0) {

		if (matrix[i][j] == key) {

			// If key is found
			printf("Element fouund at : %d %d\n",i,j);
			is_found = 1;
			break;

		} else if (matrix[i][j] > key) {

			// If current element is greater than the key, then the current column can be skipped and move to previous column
			j--;

		} else {

			// If current element is lesser than th key, then the current row can be skipped and move to next row
			i++;

		}

	}

	// If the element is not found
	if (!is_found)
		printf("Element not found\n");
}


int main() {

	// Take in input
	int n, m,ele,i,j;
	printf("Enter the dimensions of the array: ");
	scanf("%d %d",&n,&m);
  
	printf("Enter the element to be searched: ");
	scanf("%d",&ele);
	
  // Declare 2D Matrix dynamically
	int** a = (int **)malloc(n*sizeof(int *));
	for (i = 0; i < n; i++) {
		a[i] = (int *)malloc(m*sizeof(int));
	}
  printf("Enter the elements of the array: \n";
	// Take matrix input
	for (i = 0; i < n; i++) {
		for (j = 0; j < m; j++) {
			scanf("%d",&a[i][j]);
		}
	}
  //Function to search the element
	staircase_search(a, n, m, ele);
}

/*
Input:
Enter the dimensions of the array: 5 5
Enter the element to be searched: 20
Enter the elements of te array: 
1 4 8 10 45
2 5 9 15 55
6 16 18 20 12
7 14 21 27 11
11 17 19 23 19
Output:
2 3
*/

/*
Time Complexity: O(max(n, m))
Auxiliary Space: O(1)
*/
